File: language-analysis/4.Code-Smell-Refactoring/clean_implementation.py
def clean_implementation(account, action, amount=0):
    balance = account.get("balance", 0)

    if action == "balance":
        return {"status": "success", "balance": balance}

    if action == "deposit":
        if amount <= 0:
            raise ValueError("Deposit amount must be greater than zero.")
        balance += amount
        account["balance"] = balance
        return {"status": "success", "balance": balance}

    if action == "withdraw":
        if amount <= 0:
            raise ValueError("Withdrawal amount must be greater than zero.")
        if amount > balance:
            raise ValueError("Insufficient funds.")
        balance -= amount
        account["balance"] = balance
        return {"status": "success", "balance": balance}

    raise ValueError("Unknown action.")


account = {"balance": 1000}

print("\nRefactored implementation")
print(clean_implementation(account, "balance"))
print(clean_implementation(account, "deposit", 500))
print(clean_implementation(account, "withdraw", 200))

File: language-analysis/4.Code-Smell-Refactoring/clean_implementation.js
function cleanImplementation(account, action, amount = 0) {
  const balance = account.balance ?? 0;

  if (action === "balance") {
    return { status: "success", balance };
  }

  if (action === "deposit") {
    if (!Number.isInteger(amount)) {
      throw new Error("Deposit amount must be an integer.");
    }

    if (amount <= 0) {
      throw new Error("Deposit amount must be greater than zero.");
    }

    const updatedBalance = balance + amount;
    account.balance = updatedBalance;
    return { status: "success", balance: updatedBalance };
  }

  if (action === "withdraw") {
    if (!Number.isInteger(amount)) {
      throw new Error("Withdrawal amount must be an integer.");
    }

    if (amount <= 0) {
      throw new Error("Withdrawal amount must be greater than zero.");
    }

    if (amount > balance) {
      throw new Error("Insufficient funds.");
    }

    const updatedBalance = balance - amount;
    account.balance = updatedBalance;
    return { status: "success", balance: updatedBalance };
  }
}

function runAction(account, action, amount = 0) {
  try {
    console.log(cleanImplementation(account, action, amount));
  } catch (error) {
    console.log(error.message);
  }
}

const account = { balance: 1000 };

console.log("\nRefactored implementation");
runAction(account, "balance");
runAction(account, "deposit", 500);
runAction(account, "withdraw", 111200);
runAction(account, "deposit", "sss");
runAction(account, "withdraw", 10.5);
runAction(account, "deposit", 0);
runAction(account, "withdraw", 0);
